### ISSUE TITLE
Official OpenBSD support: last steps

### ISSUE THREAD

**Issue by Blaisorblade** [commercialhaskell/stack#3313]
After #3184 is merged, there might be something left for OpenBSD support. Since navigating #416 is by now impossible, I've chosen to close it. Let me paste here this TODO list from @borsboom (https://github.com/commercialhaskell/stack/issues/416#issuecomment-220801230):

> Steps remaining for official OpenBSD support:
- [x] Have `stack setup` use bzcat/xzcat to uncompress GHC bindist instead of relying on `tar` to do it (see also https://github.com/commercialhaskell/stack/issues/2283)
- [x] Await a Stackage LTS snapshot that uses GHC 8.0.x and update the Stack build process (and integration tests) to use it
- [ ] Ensure integration tests pass
- [ ] Integrate into release process (I've already added a starter Vagrantfile in `etc/vagrant/openbsd-5.9-amd64/Vagrantfile`)
- [x] Perhaps remove old not-quite-working OpenBSD bindists from [stack-setup-2.yaml](https://github.com/fpco/stackage-content/blob/master/stack/stack-setup-2.yaml), to avoid confusion.

**Comment by Blaisorblade** [comment]
Also: just noticed that @bonds [tested a bindist for GHC 8.0.2 on OpenBSD 6.1](https://github.com/commercialhaskell/stack/issues/2822#issuecomment-319147380), from
https://downloads.haskell.org/~ghc/8.0.2/ghc-8.0.2-x86_64-openbsd61-openbsd.tar.xz.

I'm not adding that to the current stack-setup PR to avoid delaying it further—and also because ping-pong on that would take days.

**Comment by bonds** [comment]
Installing 8.2.1 works, installing 8.0.2 does not:
````
~ ❯❯❯ stack setup 8.0.2                                                     Mon 2017-08-07 09:14:50 -0700
No information found for ghc-8.0.2.
Supported versions for OS key 'openbsd64-maj6-min1': GhcVersion 8.2.1
~  [1] ❯❯❯ stack setup 8.2.1                                                Mon 2017-08-07 09:14:56 -0700
Preparing to install GHC (maj6-min1) to an isolated location.
This will not interfere with any system-level installation.
Already downloaded.
Installed GHC.
stack will use a sandboxed GHC it installed
For more information on paths, see 'stack path' and 'stack exec env'
To use this GHC and packages outside of a project, consider using:
stack ghc, stack ghci, stack runghc, or stack exec
~ ❯❯❯ stack --version                                                       Mon 2017-08-07 10:14:52 -0700
Version 1.5.1, Git revision 1f5d34723996d7ce1cd2f35b5950d43d8cbdf555 (dirty) (5053 commits) x86_64 hpack-0.17.1

````

**Comment by Blaisorblade** [comment]
> ~ ❯❯❯ stack setup 8.0.2                                                     Mon 2017-08-07 09:14:50 -0700
> No information found for ghc-8.0.2.
> Supported versions for OS key 'openbsd64-maj6-min1': GhcVersion 8.2.1

Wow, the "Supported versions" bit is not at all my doing, pretty cool!
I haven't uploaded 8.0.2 yet, sorry if it wasn't clear (life intervening).

**Comment by Blaisorblade** [comment]
Wait I'm confused — https://downloads.haskell.org/~ghc/8.0.2/ghc-8.0.2-x86_64-openbsd61-openbsd.tar.xz gives 404 and there's no bindist for that combo [in that folder](https://downloads.haskell.org/~ghc/8.0.2/).

From https://github.com/commercialhaskell/stack/issues/2822#issuecomment-319147380 I assumed such a bindist existed and I just needed to upload it?

**Comment by bonds** [comment]
Hmm, yah that URL gives a 404. I'll try to see if the bindist that's in that folder works on OpenBSD 6.1, maybe I copy-pasted the wrong URL, but my memory is that I had 8.0.2 working.

**Comment by bonds** [comment]
Ok, so the default bindist for 8.0.2 for openbsd seems to run on OpenBSD 6.1:

````
~  [1] ❯❯❯ ulimit -d 2000000; nice stack setup 8.0.2                        Thu 2017-08-31 14:26:06 -0700
Preparing to install GHC to an isolated location.
This will not interfere with any system-level installation.
Already downloaded.
Installed GHC.
stack will use a locally installed GHC
For more information on paths, see 'stack path' and 'stack exec env'
To use this GHC and packages outside of a project, consider using:
stack ghc, stack ghci, stack runghc, or stack exec
~ ❯❯❯ stack --resolver ghc-8.0.2 ghc -- --version                           Thu 2017-08-31 16:01:24 -0700
Run from outside a project, using implicit global project config
Using resolver: ghc-8.0.2 specified on command line
The Glorious Glasgow Haskell Compilation System, version 8.0.2
~ ❯❯❯ stack --version                                                       Thu 2017-08-31 16:01:38 -0700
Version 1.1.2 x86_64 hpack-0.14.1
````
So, I think that means the fixed URL for the bindist in that comment is this:
https://downloads.haskell.org/~ghc/8.0.2/ghc-8.0.2-x86_64-unknown-openbsd.tar.xz and if you rename that to have the openbsd61 version number in it, it could work with the new version scheme.

**Comment by bonds** [comment]
Default bindist for 8.2.1 seems to target OpenBSD 6.0 instead of OpenBSD 6.1:
````
~ ❯❯❯ ulimit -d 2000000; nice stack setup 8.2.1                             Thu 2017-08-31 16:03:58 -0700
Preparing to install GHC to an isolated location.
This will not interfere with any system-level installation.
Already downloaded.
Running /usr/local/bin/gmake install in directory /tmp/stack-setup99428/ghc-8.2.1/ exited with ExitFailure 2

/usr/local/bin/gmake --no-print-directory -f ghc.mk install BINDIST=YES NO_INCLUDE_DEPS=YES
"rm" -f utils/ghc-pkg/dist-install/build/Version.hs
echo "module Version where"                    >> utils/ghc-pkg/dist-install/build/Version.hs
echo "version, targetOS, targetARCH :: String" >> utils/ghc-pkg/dist-install/build/Version.hs
echo "version    = \"8.2.1\""      >> utils/ghc-pkg/dist-install/build/Version.hs
echo "targetOS   = \"openbsd\""        >> utils/ghc-pkg/dist-install/build/Version.hs
echo "targetARCH = \"x86_64\""      >> utils/ghc-pkg/dist-install/build/Version.hs
"rm" -f utils/ghc-pkg/dist/build/Version.hs
echo "module Version where"                    >> utils/ghc-pkg/dist/build/Version.hs
echo "version, targetOS, targetARCH :: String" >> utils/ghc-pkg/dist/build/Version.hs
echo "version    = \"8.2.1\""      >> utils/ghc-pkg/dist/build/Version.hs
echo "targetOS   = \"openbsd\""        >> utils/ghc-pkg/dist/build/Version.hs
echo "targetARCH = \"x86_64\""      >> utils/ghc-pkg/dist/build/Version.hs
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"
"rm" -f                                         "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hp2ps"
create () { touch "$1" && chmod 755 "$1" ; } && create                                           "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hp2ps"
echo '#!/bin/sh'                                       >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hp2ps"
echo 'exedir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/bin"'                    >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hp2ps"
echo 'exeprog="hp2ps"'                          >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hp2ps"
echo 'executablename="$exedir/$exeprog"'           >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hp2ps"
echo 'datadir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/share"'                             >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hp2ps"
echo 'bindir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hp2ps"
echo 'topdir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hp2ps"
cat utils/hp2ps/hp2ps.wrapper                         >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hp2ps"
chmod +x                                         "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hp2ps"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"
"rm" -f                                 "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghci-8.2.1"
create () { touch "$1" && chmod 755 "$1" ; } && create                                   "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghci-8.2.1"
echo '#!/bin/sh'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghci-8.2.1"
echo 'exec "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-8.2.1" --interactive "$@"' >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghci-8.2.1"
chmod +x                                 "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghci-8.2.1"
"rm" -f "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghci"
ln -s ghci-8.2.1 "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghci"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/include"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/include/." && /usr/bin/install -c -m 644  includes/./*.h "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/include/./" &&   /usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/include/rts" && /usr/bin/install -c -m 644  includes/rts/*.h "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/include/rts/" &&   /usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/include/rts/prof" && /usr/bin/install -c -m 644  includes/rts/prof/*.h "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/include/rts/prof/" &&   /usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/include/rts/storage" && /usr/bin/install -c -m 644  includes/rts/storage/*.h "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/include/rts/storage/" &&   /usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/include/stg" && /usr/bin/install -c -m 644  includes/stg/*.h "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/include/stg/" &&  true
/usr/bin/install -c -m 644  includes/ghcautoconf.h includes/ghcplatform.h includes/ghcversion.h includes/dist-derivedconstants/header/DerivedConstants.h "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/include/"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"
"rm" -f                                         "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/haddock-ghc-8.2.1"
create () { touch "$1" && chmod 755 "$1" ; } && create                                           "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/haddock-ghc-8.2.1"
echo '#!/bin/sh'                                       >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/haddock-ghc-8.2.1"
echo 'exedir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/bin"'                    >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/haddock-ghc-8.2.1"
echo 'exeprog="haddock"'                          >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/haddock-ghc-8.2.1"
echo 'executablename="$exedir/$exeprog"'           >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/haddock-ghc-8.2.1"
echo 'datadir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/share"'                             >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/haddock-ghc-8.2.1"
echo 'bindir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/haddock-ghc-8.2.1"
echo 'topdir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/haddock-ghc-8.2.1"
cat utils/haddock/haddock.wrapper                         >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/haddock-ghc-8.2.1"
chmod +x                                         "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/haddock-ghc-8.2.1"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/" 
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/Classic.theme/"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/Ocean.std-theme/"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/latex/"
/usr/bin/install -c -m 644  utils/haddock/haddock-api/resources/html/haddock-util.js "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/"
/usr/bin/install -c -m 644  utils/haddock/haddock-api/resources/html/Classic.theme/haskell_icon.gif "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/Classic.theme/"
/usr/bin/install -c -m 644  utils/haddock/haddock-api/resources/html/Classic.theme/minus.gif "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/Classic.theme/"
/usr/bin/install -c -m 644  utils/haddock/haddock-api/resources/html/Classic.theme/plus.gif "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/Classic.theme/"
/usr/bin/install -c -m 644  utils/haddock/haddock-api/resources/html/Classic.theme/xhaddock.css "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/Classic.theme/"
/usr/bin/install -c -m 644  utils/haddock/haddock-api/resources/html/Ocean.std-theme/hslogo-16.png "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/Ocean.std-theme/"
/usr/bin/install -c -m 644  utils/haddock/haddock-api/resources/html/Ocean.std-theme/minus.gif "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/Ocean.std-theme/"
/usr/bin/install -c -m 644  utils/haddock/haddock-api/resources/html/Ocean.std-theme/ocean.css "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/Ocean.std-theme/"
/usr/bin/install -c -m 644  utils/haddock/haddock-api/resources/html/Ocean.std-theme/plus.gif "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/Ocean.std-theme/"
/usr/bin/install -c -m 644  utils/haddock/haddock-api/resources/html/Ocean.std-theme/synopsis.png "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/Ocean.std-theme/"
/usr/bin/install -c -m 644  utils/haddock/haddock-api/resources/html/solarized.css "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/"
/usr/bin/install -c -m 644  utils/haddock/haddock-api/resources/html/highlight.js "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/html/"
/usr/bin/install -c -m 644  utils/haddock/haddock-api/resources/latex/haddock.sty "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/latex/"
"rm" -f "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/haddock"
ln -s haddock-ghc-8.2.1 "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/haddock"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"
"rm" -f                                         "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hsc2hs"
create () { touch "$1" && chmod 755 "$1" ; } && create                                           "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hsc2hs"
echo '#!/bin/sh'                                       >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hsc2hs"
echo 'exedir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/bin"'                    >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hsc2hs"
echo 'exeprog="hsc2hs"'                          >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hsc2hs"
echo 'executablename="$exedir/$exeprog"'           >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hsc2hs"
echo 'datadir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/share"'                             >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hsc2hs"
echo 'bindir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hsc2hs"
echo 'topdir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hsc2hs"
echo 'HSC2HS_EXTRA="--cflag=-std=gnu99 --cflag=-fno-stack-protector --lflag=-Wl,-z,wxneeded"' >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hsc2hs"
cat utils/hsc2hs/hsc2hs.wrapper                         >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hsc2hs"
chmod +x                                         "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hsc2hs"
/usr/bin/install -c -m 644  utils/hsc2hs/template-hsc.h "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"
"rm" -f                                         "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-pkg-8.2.1"
create () { touch "$1" && chmod 755 "$1" ; } && create                                           "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-pkg-8.2.1"
echo '#!/bin/sh'                                       >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-pkg-8.2.1"
echo 'exedir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/bin"'                    >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-pkg-8.2.1"
echo 'exeprog="ghc-pkg"'                          >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-pkg-8.2.1"
echo 'executablename="$exedir/$exeprog"'           >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-pkg-8.2.1"
echo 'datadir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/share"'                             >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-pkg-8.2.1"
echo 'bindir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-pkg-8.2.1"
echo 'topdir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-pkg-8.2.1"
cat utils/ghc-pkg/ghc-pkg.wrapper                         >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-pkg-8.2.1"
chmod +x                                         "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-pkg-8.2.1"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"
"rm" -f "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-pkg"
ln -s ghc-pkg-8.2.1 "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-pkg"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"
"rm" -f                                         "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hpc"
create () { touch "$1" && chmod 755 "$1" ; } && create                                           "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hpc"
echo '#!/bin/sh'                                       >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hpc"
echo 'exedir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/bin"'                    >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hpc"
echo 'exeprog="hpc"'                          >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hpc"
echo 'executablename="$exedir/$exeprog"'           >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hpc"
echo 'datadir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/share"'                             >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hpc"
echo 'bindir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hpc"
echo 'topdir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hpc"
cat utils/hpc/hpc.wrapper                         >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hpc"
chmod +x                                         "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/hpc"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"
"rm" -f                                         "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runghc-8.2.1"
create () { touch "$1" && chmod 755 "$1" ; } && create                                           "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runghc-8.2.1"
echo '#!/bin/sh'                                       >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runghc-8.2.1"
echo 'exedir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/bin"'                    >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runghc-8.2.1"
echo 'exeprog="runghc"'                          >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runghc-8.2.1"
echo 'executablename="$exedir/$exeprog"'           >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runghc-8.2.1"
echo 'datadir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/share"'                             >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runghc-8.2.1"
echo 'bindir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runghc-8.2.1"
echo 'topdir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runghc-8.2.1"
echo 'ghcprog="ghc-8.2.1"' >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runghc-8.2.1"
cat utils/runghc/runghc.wrapper                         >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runghc-8.2.1"
chmod +x                                         "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runghc-8.2.1"
"rm" -f "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runhaskell"
ln -s runghc "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runhaskell"
"rm" -f "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runghc"
ln -s runghc-8.2.1 "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/runghc"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"
"rm" -f                                         "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-8.2.1"
create () { touch "$1" && chmod 755 "$1" ; } && create                                           "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-8.2.1"
echo '#!/bin/sh'                                       >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-8.2.1"
echo 'exedir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/bin"'                    >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-8.2.1"
echo 'exeprog="ghc-stage2"'                          >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-8.2.1"
echo 'executablename="$exedir/$exeprog"'           >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-8.2.1"
echo 'datadir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/share"'                             >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-8.2.1"
echo 'bindir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin"'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-8.2.1"
echo 'topdir="/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"'                               >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-8.2.1"
echo 'executablename="$exedir/ghc"' >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-8.2.1"
cat ghc/ghc.wrapper                         >> "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-8.2.1"
chmod +x                                         "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc-8.2.1"
"rm" -f "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc"
ln -s ghc-8.2.1 "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/bin/ghc"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/share/man"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/share/man/man1"
/usr/bin/install -c -m 644  docs/users_guide/build-man/ghc.1 "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/share/man/man1"
#  driver/ghc-usage.txt driver/ghci-usage.txt includes/dist-derivedconstants/header/platformConstants settings = libraries to install
#  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1" = directory to install to
#
# The .dll case calls STRIP_CMD explicitly, instead of `install -s`, because
# on Win64, "install -s" calls a strip that doesn't understand 64bit binaries.
# For some reason, this means the DLLs end up non-executable, which means
# executables that use them just segfault.
/usr/bin/install -c -m 755 -d  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"
for i in  driver/ghc-usage.txt driver/ghci-usage.txt includes/dist-derivedconstants/header/platformConstants settings; do case $i in *.a) /usr/bin/install -c -m 644  $i  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"; true  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"/`basename $i` ;; *.dll) /usr/bin/install -c -m 755  $i  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1" ; strip  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"/`basename $i` ;; *.so) /usr/bin/install -c -m 755  $i  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1" ;; *.dylib) /usr/bin/install -c -m 755  $i  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1";; *) /usr/bin/install -c -m 644  $i  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"; esac; done
gcc -E -undef -traditional -P -DINSTALLING -DLIB_DIR='"/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"' -DINCLUDE_DIR='"/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/include"' -DFFI_INCLUDE_DIR=/usr/local/include -DFFI_LIB_DIR=/usr/local/lib '-DFFI_LIB=' -x c -Iincludes -Iincludes/dist -Iincludes/dist-derivedconstants/header -Iincludes/dist-ghcconstants/header rts/package.conf.in -o rts/dist/package.conf.install.raw
grep -v '^#pragma GCC' rts/dist/package.conf.install.raw | sed -e 's/""//g' -e 's/:[    ]*,/: /g' >rts/dist/package.conf.install
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/bin"
for i in utils/unlit/dist/build/tmp/unlit utils/hp2ps/dist/build/tmp/hp2ps utils/hp2ps/dist/build/tmp/hp2ps utils/haddock/dist/build/tmp/haddock utils/haddock/dist/build/tmp/haddock utils/hsc2hs/dist-install/build/tmp/hsc2hs utils/hsc2hs/dist-install/build/tmp/hsc2hs utils/ghc-pkg/dist-install/build/tmp/ghc-pkg utils/ghc-pkg/dist-install/build/tmp/ghc-pkg utils/hpc/dist-install/build/tmp/hpc utils/hpc/dist-install/build/tmp/hpc utils/runghc/dist-install/build/tmp/runghc utils/runghc/dist-install/build/tmp/runghc ghc/stage2/build/tmp/ghc-stage2 ghc/stage2/build/tmp/ghc-stage2 iserv/stage2/build/tmp/ghc-iserv iserv/stage2_p/build/tmp/ghc-iserv-prof iserv/stage2_dyn/build/tmp/ghc-iserv-dyn; do \
        /usr/bin/install -c -m 755  $i "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/bin"; \
done
"mv" "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/bin/ghc-stage2" "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/bin/ghc"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1"
"rm" -rf "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/package.conf.d"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/package.conf.d"
/usr/bin/install -c -m 755 -d "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/rts"
#  rts/dist/build/libHSrts.a rts/dist/build/libHSrts_p.a rts/dist/build/libHSrts-ghc8.2.1.so rts/dist/build/libHSrts_l.a rts/dist/build/libHSrts_debug.a rts/dist/build/libHSrts_thr.a rts/dist/build/libHSrts_thr_debug.a rts/dist/build/libHSrts_thr_l.a rts/dist/build/libHSrts_thr_p.a rts/dist/build/libHSrts_debug-ghc8.2.1.so rts/dist/build/libHSrts_thr-ghc8.2.1.so rts/dist/build/libHSrts_thr_debug-ghc8.2.1.so rts/dist/build/libHSrts_l-ghc8.2.1.so rts/dist/build/libHSrts_thr_l-ghc8.2.1.so = libraries to install
#  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/rts" = directory to install to
#
# The .dll case calls STRIP_CMD explicitly, instead of `install -s`, because
# on Win64, "install -s" calls a strip that doesn't understand 64bit binaries.
# For some reason, this means the DLLs end up non-executable, which means
# executables that use them just segfault.
/usr/bin/install -c -m 755 -d  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/rts"
for i in  rts/dist/build/libHSrts.a rts/dist/build/libHSrts_p.a rts/dist/build/libHSrts-ghc8.2.1.so rts/dist/build/libHSrts_l.a rts/dist/build/libHSrts_debug.a rts/dist/build/libHSrts_thr.a rts/dist/build/libHSrts_thr_debug.a rts/dist/build/libHSrts_thr_l.a rts/dist/build/libHSrts_thr_p.a rts/dist/build/libHSrts_debug-ghc8.2.1.so rts/dist/build/libHSrts_thr-ghc8.2.1.so rts/dist/build/libHSrts_thr_debug-ghc8.2.1.so rts/dist/build/libHSrts_l-ghc8.2.1.so rts/dist/build/libHSrts_thr_l-ghc8.2.1.so; do case $i in *.a) /usr/bin/install -c -m 644  $i  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/rts"; true  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/rts"/`basename $i` ;; *.dll) /usr/bin/install -c -m 755  $i  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/rts" ; strip  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/rts"/`basename $i` ;; *.so) /usr/bin/install -c -m 755  $i  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/rts" ;; *.dylib) /usr/bin/install -c -m 755  $i  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/rts";; *) /usr/bin/install -c -m 644  $i  "/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1/rts"; esac; done
"utils/ghc-cabal/dist-install/build/tmp/ghc-cabal-bindist" copy libraries/ghc-prim dist-install "strip" '' '/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1' '/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/lib/ghc-8.2.1' '/home/scott/.stack/programs/x86_64-openbsd/ghc-8.2.1/share/doc/ghc-8.2.1/html/libraries' 'v p dyn'

ld.so: ghc-cabal: can't load library 'libgmp.so.9.0'
Killed
gmake[1]: *** [ghc.mk:991: install_packages] Error 137
gmake: *** [Makefile:51: install] Error 2

Installing GHC ...⏎                                                                                      
~  [1] ❯❯❯ locate libgmp.so                                                 Thu 2017-08-31 16:08:13 -0700
/usr/local/lib/libgmp.so.10.0
~
````
So for that one you need a custom stack.yaml to point it at the bindist for 6.1 instead of 6.0, if you're on 6.1.

**Comment by bonds** [comment]
Does that help @Blaisorblade ?

**Comment by Blaisorblade** [comment]
@bonds Thanks for the ping, sorry for my absence (been busy). I'll try tomorrow—I need the time to reparse the comments.

TODOs:
- [ ] upload https://downloads.haskell.org/~ghc/8.0.2/ghc-8.0.2-x86_64-unknown-openbsd.tar.xz as GHC 8.0.2 for openbsd61, and add it to stack-setup.

On your later comment:
> Default bindist for 8.2.1 seems to target OpenBSD 6.0 instead of OpenBSD 6.1:

Yes, that's visible in the default URL.
https://github.com/fpco/stackage-content/blob/422a85942bdb65eb7cc766585c0a0da1d4f7174f/stack/stack-setup-2.yaml#L627-L628

> [...]
> So for that one you need a custom stack.yaml to point it at the bindist for 6.1 instead of 6.0, if you're on 6.1.

The interesting details are missing from that log, but can I guess that's a stack without the version autodetection PR, since it runs on 6.1 and does not use the version for 6.1 (which is specified since https://github.com/fpco/stackage-content/pull/25)? So for this I need to change nothing in stack, right? While you can update stack maybe (not sure that PR is in a release or just in master, can check tomorrow)?

- [ ] The default bindist for GHC 8.2.1 is for OpenBSD 6.0 in theory, but it doesn't actually work in practice (recheck https://github.com/fpco/stackage-content/issues/28, status of linked bugs appears unchanged). So maybe replace it by the version for OpenBSD 6.1 as default, for those stack versions still without autodetection?

**Comment by bonds** [comment]
> The interesting details are missing from that log, but can I guess that's a stack without the version autodetection PR, since it runs on 6.1 and does not use the version for 6.1 (which is specified since fpco/stackage-content#25)? 

correct, I was using "Version 1.1.2 x86_64 hpack-0.14.1" for the tests on 2017-08-31

> So for this I need to change nothing in stack, right?

correct

> While you can update stack maybe (not sure that PR is in a release or just in master, can check tomorrow)? The default bindist for GHC 8.2.1 is for OpenBSD 6.0 in theory, but it doesn't actually work in practice (recheck fpco/stackage-content#28, status of linked bugs appears unchanged). So maybe replace it by the version for OpenBSD 6.1 as default, for those stack versions still without autodetection?

I think so. I believe most of the OpenBSD user base upgrades to the lastest release, so if you update the default bindist to match the latest OpenBSD release, for folks using an older version of Stack, you'll probably help the most people. Eventually the newer version of Stack will eliminate the need to update the default binddist. OTOH, you could just be lazy and NOT update the default bindist, and let people know they need to upgrade Stack if they want 'easy mode' or alter stack.yaml to point to the right bindist if they want to use the old stack versions (maybe a wiki page entry somewhere?).

**Comment by romanzolotarev** [comment]
BTW, OpenBSD 6.2 release is coming soon: https://www.openbsd.org/62.html says `October 15, 2017`.

**Comment by borsboom** [comment]
It appears that the GHC release team has stopped building OpenBSD bindists.  There are none for GHC 8.2.2.

**Comment by 1dotd4** [comment]
**WARNING: I'm new to OpenBSD and I have done a lot of insecure things.**

```
# pkg_add ghc cabal-install
# mount -u -o wxallowed /tmp
$ cabal update
$ cabal install --force-reinstalls stack
```

it stops at compiling time:
`ghc: out of memory (requested 1048576 bytes)`

**Comment by davlse** [comment]
Hello,

Since the last comment here we've got GHC 8.6.4 in the ports tree for OpenBSD.
With that and cabal-install 2.4.0.0 I tried to bootstrap stack from Hackage as was described at the very bottom of [releases.md](https://github.com/commercialhaskell/stack/blob/master/doc/maintainers/releases.md). 

```
cabal unpack stack && \
cd stack-* && \
cabal install && \
mv ~/.cabal/bin/stack ~/.local/bin
```

It built quite a few libraries until I hit this error.

```
Building library for pantry-0.2.0.0..
[ 1 of 18] Compiling Pantry.Internal.AesonExtended ( src/Pantry/Internal/AesonExtended.hs, dist/buil
[ 2 of 18] Compiling Pantry.Internal.Companion ( src/Pantry/Internal/Companion.hs, dist/build/Pantry
[ 3 of 18] Compiling Pantry.Internal.StaticBytes ( src/Pantry/Internal/StaticBytes.hs, dist/build/Pa
[ 4 of 18] Compiling Pantry.SHA256    ( src/Pantry/SHA256.hs, dist/build/Pantry/SHA256.o )
[ 5 of 18] Compiling Pantry.Types     ( src/Pantry/Types.hs, dist/build/Pantry/Types.o )
[ 6 of 18] Compiling Pantry.SQLite    ( src/Pantry/SQLite.hs, dist/build/Pantry/SQLite.o )
[ 7 of 18] Compiling Pantry.HTTP      ( src/Pantry/HTTP.hs, dist/build/Pantry/HTTP.o )
[ 8 of 18] Compiling Hackage.Security.Client.Repository.HttpLib.HttpClient ( src/Hackage/Security/Cl

src/Hackage/Security/Client/Repository/HttpLib/HttpClient.hs:25:1: error:
    Could not load module `Hackage.Security.Util.Lens'
    it is a hidden module in the package `hackage-security-0.6.0.0'
    Use -v to see a list of the files searched for.
   |
25 | import qualified Hackage.Security.Util.Lens as Lens
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cabal: Leaving directory '/tmp/cabal-tmp-13322/pantry-0.2.0.0'
Warning: could not create symlinks in /home/davl/.cabal/bin for hpack,
haskell-mustache, example because the files exist there already and are not
managed by cabal. You can create symlinks for these executables manually if
you wish. The executable files have been installed at
/home/davl/.cabal/bin/hpack, /home/davl/.cabal/bin/haskell-mustache,
/home/davl/.cabal/bin/example
cabal: Error: some packages failed to install:
pantry-0.2.0.0-J1AYUuxFJETIh2bPE7N1rw failed during the building phase. The
exception was:
ExitFailure 1
stack-2.1.3.1-76CuvunVxcmFzSS2RmwzVd depends on stack-2.1.3.1 which failed to
install.
```
I am a bit confused about the warning at the end as the .cabal/bin directory didn't exist when I started so cabal must have put those binaries there even if it doesn't think so...

**Comment by borsboom** [comment]
Looks like the recently released hackage-security-0.6.0.0 has some breaking changes.  Try adding a `< 0.6.0.0` constraint in `stack.cabal`.

**Comment by davlse** [comment]
Thanks that helped build all the dependencies. I still have problems building the stack package.

There was a simple ambiguous import in src/Stack/Package.hs and then some more annoying type errors in src/Stack/Setup.hs

```
[ 88 of 110] Compiling Stack.Setup      ( src/Stack/Setup.hs, dist/build/Stack/Setup.o )

src/Stack/Setup.hs:879:29: error:
    * Could not deduce (MonadThrow m0)
        arising from a use of `fileExtension'
      from the context: HasConfig env
        bound by the type signature for:
                   buildGhcFromSource :: forall env.
                                         (HasTerm env, HasProcessContext env, HasConfig env) =>
                                         Memoized SetupInfo
                                         -> [Tool]
                                         -> CompilerRepository
                                         -> Text
                                         -> Text
                                         -> RIO env (Tool, CompilerBuild)
        at src/Stack/Setup.hs:(826,1)-(831,35)
      The type variable `m0' is ambiguous
      These potential instances exist:
        instance (e ~ SomeException) => MonadThrow (Either e)
          -- Defined in `Control.Monad.Catch'
        instance MonadThrow STM -- Defined in `Control.Monad.Catch'
        instance MonadThrow (ST s) -- Defined in `Control.Monad.Catch'
        ...plus 7 others
        ...plus 17 instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
    * In the first argument of `(==)', namely
        `fileExtension (filename p)'
      In the second argument of `(&&)', namely
        `fileExtension (filename p) == ".xz"'
      In the expression:
        "ghc-" `isPrefixOf` (toFilePath (filename p))
          && fileExtension (filename p) == ".xz"
    |
879 |                          && fileExtension (filename p) == ".xz"
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Stack/Setup.hs:879:29: error:
    * Could not deduce (Eq (m0 String)) arising from a use of `=='
      from the context: HasConfig env
        bound by the type signature for:
                   buildGhcFromSource :: forall env.
                                         (HasTerm env, HasProcessContext env, HasConfig env) =>
                                         Memoized SetupInfo
                                         -> [Tool]
                                         -> CompilerRepository
                                         -> Text
                                         -> Text
                                         -> RIO env (Tool, CompilerBuild)
        at src/Stack/Setup.hs:(826,1)-(831,35)
      The type variable `m0' is ambiguous
      These potential instances exist:
        instance Eq a => Eq (Result a)
          -- Defined in `aeson-1.4.6.0:Data.Aeson.Types.Internal'
        instance Eq (Async a) -- Defined in `Control.Concurrent.Async'
        instance (Eq a, Eq b) => Eq (Either a b)
          -- Defined in `Data.Either'
        ...plus 43 others
        ...plus 117 instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
    * In the second argument of `(&&)', namely
        `fileExtension (filename p) == ".xz"'
      In the expression:
        "ghc-" `isPrefixOf` (toFilePath (filename p))
          && fileExtension (filename p) == ".xz"
      In an equation for `isBindist':
          isBindist p
            = "ghc-" `isPrefixOf` (toFilePath (filename p))
                && fileExtension (filename p) == ".xz"
    |
879 |                          && fileExtension (filename p) == ".xz"
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Stack/Setup.hs:879:59: error:
    * Could not deduce (IsString (m0 String))
        arising from the literal `".xz"'
      from the context: HasConfig env
        bound by the type signature for:
                   buildGhcFromSource :: forall env.
                                         (HasTerm env, HasProcessContext env, HasConfig env) =>
                                         Memoized SetupInfo
                                         -> [Tool]
                                         -> CompilerRepository
                                         -> Text
                                         -> Text
                                         -> RIO env (Tool, CompilerBuild)
        at src/Stack/Setup.hs:(826,1)-(831,35)
      The type variable `m0' is ambiguous
      These potential instances exist:
        instance IsString a => IsString (Identity a)
          -- Defined in `Data.String'
        instance (a ~ Char) => IsString (Seq a)
          -- Defined in `Data.Sequence.Internal'
        instance (stdin ~ (), stdout ~ (), stderr ~ ()) =>
                 IsString (ProcessConfig stdin stdout stderr)
          -- Defined in `System.Process.Typed'
        ...plus two others
        ...plus 9 instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
    * In the second argument of `(==)', namely `".xz"'
      In the second argument of `(&&)', namely
        `fileExtension (filename p) == ".xz"'
      In the expression:
        "ghc-" `isPrefixOf` (toFilePath (filename p))
          && fileExtension (filename p) == ".xz"
    |
879 |                          && fileExtension (filename p) == ".xz"
    |                                                           ^^^^^
cabal: Leaving directory '.'
cabal: Error: some packages failed to install:
stack-2.1.3.1-JqvuJLq3K0f99T5bziO1Js failed during the building phase. The
exception was:
ExitFailure 1
```

Do you think more library changes?

**Comment by Blaisorblade** [comment]
fileExtension used to return a String, but changed in 0.7.0 (https://hackage.haskell.org/package/path-0.7.0/changelog). Try adding a path < 0.7.0 constraint in stack.cabal.

**Comment by davlse** [comment]
Those constraints was the only ones needed. After that it produced the stack binary as expected.

I built the purescript compiler as a test. I did have to set both "system-ghc: true" and "skip-ghc-check: true" in config.yaml and bump the limits in /etc/login.conf quite high (the Cabal package used in excess of 2GB when building). Other than that it seems to work fine.

Do you have anything else I should try?

**Comment by 1dotd4** [comment]
Was trying to follow @davlse's guide, so started again from ghc and cabal-install, openbsd's packages (6.7-current).

After fetching with cabal the stack package and changed the two requirements[1] and running `cabal install --force-reinstalls` I get the following logs[2].

```
src/System/Time/Extra.hs:16:1: error:
    Could not find module `System.Clock'
    There are files missing in the `clock-0.8' package,
    try running 'ghc-pkg check'.
    Use -v to see a list of the files searched for.
   |
16 | import System.Clock
   | ^^^
```

What am I missing?

[1] hackage-security < 0.6.0.0, path <0.7.0
[2] https://gist.github.com/1dotd4/dacd91716dc8b4bce7f2be739457f5f9

**Comment by 1dotd4** [comment]
Uh, looks like I needed to uninstall newer packages first 🤦 (sorry). Then worked, running `cabal install`.

So I guess last step is contacting an openbsd maintainer to pack this up. (Right?)

**Comment by 1dotd4** [comment]
I was trying to follow @davlse's guide, so started again from `ghc` and `cabal-install`, openbsd's packages (I'm in `6.7-current`). After fetching with cabal the stack package (`cabal unpack stack`) and changed the two requirements<sup>1</sup> and running `cabal install` I get the following logs<sup>2</sup>.

Looks like there are _ambiguous occurrences_ between `Stack.Prelude` and `Stack.Types.Resolver`. But I have no idea of what I should tweak to get finish the compilation.

<sup>1</sup> hackage-security < 0.6.0.0, path <0.7.0
<sup>2</sup> https://gist.github.com/1dotd4/dacd91716dc8b4bce7f2be739457f5f9

**Comment by eyeinsky** [comment]
I've succeeded building stack on OpenBSD 6.8, this is what I did:
1. set `datasize-cur` in `/etc/login.conf` to infinity (I made myself part of the staff group and changed the staff section in the login.conf)
1. `doas pkg_add ghc cabal-install` which installs ghc-8.6.4 and cabal-install 2.4.0.0
1. `cabal new-update`
1. `mkdir -p /home/markus/.cabal/store/ghc-8.6.4/package.db` (not sure why this is required, but it helped)
1. `cabal new-install cabal-install`, after which I had cabal-intall version 3.2.0.0
1. Add `$HOME/.cabal/bin` to `PATH` in `~/.profile`
1. `cabal unpack stack` and add `{-# LANGUAGE DataKinds #-}` to src/Stack/Storage/User.hs and src/Stack/Storage/Project.hs (I made an issue here https://github.com/commercialhaskell/stack/issues/5438 but perhaps that should be closed in favor of the current thread)
1. `cabal install` in the stack folder, and now it succeeded

Now that I tried to build a project with stack I got this:
`Unable to find installation URLs for OS key: openbsd64-maj6-min8`

This error goes away with `stack build --resolver ghc-8.6.4 --system-ghc` (but then i get a `Plan construction failed.`).

But in short, what are the last steps the "official OpenBSD support" require -- so that a simple `stack build` with any ghc version would would work? Or can we just support the ones that have a prebuilt ghc available?

**Comment by ketzacoatl** [comment]
Right now, for someone to use `stack` on OpenBSD, they need to search around and find this post, then read the latest comments to figure out what the current method is. Ideally, we'd have a section or page in the docs that is responsible for containing "the current method".

Even if the one-line installer script works for OpenBSD, it would be helpful to add notes such as pinning GHC to the system install, or the fact that not all LTS will work (and a user may need to hunt around to find the right one).

@borsboom, maybe you can recommend a location for such a section? Maybe we could add a section for OpenBSD in the install docs?

**Comment by distefam** [comment]
I can confirm what @eyeinsky has said. I followed the instructions in this thread and got one step further than he reported by following the suggested instructions printed above the `plan construction failed` error: to manually add a bunch of dependencies to the `package.yaml` file. However, this too failed.

**Comment by eyeinsky** [comment]
I think this issue hinges on having official ghc builds available for OpenBSD, but the latest official one available is [8.2.1](https://www.haskell.org/ghc/download_ghc_8_2_1.html#openbsd_61_x86_64), and the latest that can be installed with `pkg_add` (and thus used via ` --system-ghc`) is 8.6.4.

**Comment by adamsteen** [comment]
just wanted to update this issues for people who find it.

i was able to a `cabal update` and `cabal install stack`

**Comment by adamsteen** [comment]
Updated a year later, as per @eyeinsky 

1. set `datasize-cur` in `/etc/login.conf` to infinity (I made myself part of the staff group and changed the staff section in the login.conf)
2. `doas pkg_add ghc cabal-install` which installs ghc 8.10.6 and cabal-install 3.6.2.0
3. `cabal unpack stack`
4. `cabal install` in the stack folder, and now it fails with
```
Configuring library for pantry-0.5.2.3..
Preprocessing library for pantry-0.5.2.3..
Building library for pantry-0.5.2.3..
[ 1 of 19] Compiling Pantry.Internal.AesonExtended ( src/Pantry/Internal/AesonExtended.hs, dist/build/Pantry/Internal/AesonExtended.o, dist/build/Pantry/Internal/AesonExtended.dyn_o )

src/Pantry/Internal/AesonExtended.hs:48:88: error:
    • Couldn't match expected type ‘Key’ with actual type ‘Text’
    • In the second argument of ‘(A..:)’, namely ‘p’
      In the second argument of ‘modifyFailure’, namely ‘(o A..: p)’
      In the expression:
        modifyFailure
          (("failed to parse field '" <> unpack p <> "': ") <>) (o A..: p)
   |
48 | (.:) o p = modifyFailure (("failed to parse field '" <> unpack p <> "': ") <>) (o A..: p)
   |                                                                                        ^

src/Pantry/Internal/AesonExtended.hs:53:90: error:
    • Couldn't match expected type ‘Key’ with actual type ‘Text’
    • In the second argument of ‘(A..:?)’, namely ‘p’
      In the second argument of ‘modifyFailure’, namely ‘(o A..:? p)’
      In the expression:
        modifyFailure
          (("failed to parse field '" <> unpack p <> "': ") <>) (o A..:? p)
   |
53 | (.:?) o p = modifyFailure (("failed to parse field '" <> unpack p <> "': ") <>) (o A..:? p)
   |                                                                                          ^

src/Pantry/Internal/AesonExtended.hs:77:61: error:
    • Couldn't match type ‘Data.Aeson.KeyMap.KeyMap Value’
                     with ‘HashMap Text a0’
      Expected type: HashMap Text a0
        Actual type: Object
    • In the second argument of ‘HashMap.member’, namely ‘o’
      In the expression: HashMap.member x o
      In the first argument of ‘filter’, namely
        ‘(\ x -> HashMap.member x o)’
   |
77 | presentCount o ss = length . filter (\x -> HashMap.member x o) $ ss
   |                                                             ^

src/Pantry/Internal/AesonExtended.hs:124:58: error:
    • Couldn't match type ‘Data.Aeson.KeyMap.KeyMap Value’
                     with ‘HashMap Text v0’
      Expected type: HashMap Text v0
        Actual type: Object
    • In the first argument of ‘HashMap.keys’, namely ‘obj’
      In the first argument of ‘Set.fromList’, namely
        ‘(HashMap.keys obj)’
      In the first argument of ‘Set.difference’, namely
        ‘(Set.fromList (HashMap.keys obj))’
    |
124 |                              (Set.fromList (HashMap.keys obj))
    |                                                          ^^^

```

**Comment by adamsteen** [comment]
Hi All

this PR fixed my problems, https://github.com/commercialhaskell/stack/pull/5677

**Comment by adamsteen** [comment]
When is says "Ensure integration tests pass" at the top of this issue, what command do i run to test it?

i have successfully run `stack test`

with
```
Finished in 1.7492 seconds
673 examples, 0 failures

stack             > Test suite stack-test passed
Completed 15 action(s).
```

**Comment by ksilo** [comment]
One more year latter
    1. `doas pkg_add ghc cabal-install` which installs ghc 9.2.4 and cabal-install 3.6.2.0
    2. `cabal new-install cabal-install` gives cabal-install 3.10.1.0
    3. `cabal unpack stack` stack-2.11.1
    4. `cabal install` in the stack folder, and now it fails with

```
src/Options/Applicative/Complicated.hs:184:3: error:
    • Couldn't match expected type: (Maybe String,
                                     [(String, ParserInfo a)])
                  with actual type: (a3, b, c)
    • In the pattern: (groupName, cmds, subs)
      In a pattern binding: (groupName, cmds, subs) = mkCommand m
      In an equation for ‘hsubparser'’:
          hsubparser' commandMetavar m
            = mkParser d g rdr
            where
                Mod _ d g = metavar commandMetavar `mappend` m
                (groupName, cmds, subs) = mkCommand m
                rdr = CmdReader groupName cmds (fmap add_helper . subs)
                add_helper pinfo
                  = pinfo {infoParser = infoParser pinfo <**> helpOption}
    • Relevant bindings include
        m :: Mod CommandFields a
          (bound at src/Options/Applicative/Complicated.hs:181:28)
        hsubparser' :: String -> Mod CommandFields a -> Parser a
          (bound at src/Options/Applicative/Complicated.hs:181:1)
    |
184 |   (groupName, cmds, subs) = mkCommand m
    |   ^^^^^^^^^^^^^^^^^^^^^^^

src/Options/Applicative/Complicated.hs:185:9: error:
    • Couldn't match expected type: (a1 -> f0 (ParserInfo a2)) -> t
                  with actual type: OptReader a0
    • The function ‘CmdReader’ is applied to three value arguments,
        but its type ‘Maybe String
                      -> [(String, ParserInfo a0)] -> OptReader a0’
        has only two
      In the expression:
        CmdReader groupName cmds (fmap add_helper . subs)
      In an equation for ‘rdr’:
          rdr = CmdReader groupName cmds (fmap add_helper . subs)
    • Relevant bindings include
        rdr :: t (bound at src/Options/Applicative/Complicated.hs:185:3)
    |
185 |   rdr = CmdReader groupName cmds (fmap add_helper . subs)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by mpilgrem** [comment]
@ksilo, the problem is that you are using Cabal (the tool) to build Stack. If you use Stack to build Stack it should be fine. What you are experiencing is the breaking change in `optparse-applicative-0.18.0.0`, but Stack is built against an earlier version of that package - see `stack.yaml`.

**Comment by ksilo** [comment]
@mpilgrem thanks!

changed requirement for `optparse-applicative < 0.18.0.0` worked flawlessly with Cabal (the tool)

**Comment by mpilgrem** [comment]
@ksilo, that's super. Also, the `cabal.project` and `cabal.config` in the `master` branch version of Stack should now be up to date.
